ldt/src/quotient.rs

```diff
@@ -181,77 +295,215 @@ impl<F, EF, Inner: MatrixRowSlices<F>> Matrix<EF> for QuotientMatrix<F, EF, Inne
     }
 }
 
-impl<F: Field, EF: ExtensionField<F>, Inner: MatrixRowSlices<F>> MatrixRows<EF>
+impl<F: Field, EF: HasFrobenius<F>, Inner: MatrixRowSlices<F>> MatrixRows<F>
     for QuotientMatrix<F, EF, Inner>
 {
-    type Row<'a> = QuotientMatrixRow<'a, F, EF> where Inner: 'a;
+    type Row<'a> = Vec<F> where Inner: 'a;
 
     #[inline]
     fn row(&self, r: usize) -> Self::Row<'_> {
-        let num_openings = self.openings.len();
-        QuotientMatrixRow {
-            openings: &self.openings,
-            inv_denominator: &self.inv_denominators[r * num_openings..(r + 1) * num_openings],
-            inner_row: self.inner.row_slice(r),
-            opening_index: 0,
-            inner_col_index: 0,
-        }
+        compute_quotient_matrix_row(
+            self.subgroup[r],
+            &self.openings,
+            self.inv_denominators.row_slice(r),
+            self.inner.row_slice(r),
+        )
     }
-}
 
-pub struct QuotientMatrixRow<'a, F, EF> {
-    openings: &'a [Opening<EF>],
-    /// `1 / (x - opened_point)`
-    inv_denominator: &'a [EF],
-    inner_row: &'a [F],
-    opening_index: usize,
-    inner_col_index: usize,
+    #[inline]
+    fn row_vec(&self, r: usize) -> Vec<F> {
+        self.row(r)
+    }
 }
 
-impl<'a, F, EF> Iterator for QuotientMatrixRow<'a, F, EF>
-where
-    F: Field,
-    EF: ExtensionField<F>,
-{
-    type Item = EF;
+fn compute_quotient_matrix_row<F: Field, EF: HasFrobenius<F>>(
+    x: F,
+    openings: &[Opening<F, EF>],
+    m_invs: &[F],
+    inner_row: &[F],
+) -> Vec<F> {
+    let mut qp_ys: Vec<F> = Vec::with_capacity(inner_row.len() * openings.len());
 
-    #[inline]
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.inner_col_index == self.inner_row.len() {
-            self.opening_index += 1;
-            self.inner_col_index = 0;
-        }
-        if self.opening_index == self.openings.len() {
-            return None;
+    // [P(1,1,1,1), P(x,x,x,x),P(x^2,x^2,x^2,x^2),..]
+    let packed_x_pows = x
+        .powers()
+        .take(EF::D)
+        .map(|x_pow| F::Packing::from(x_pow))
+        .collect_vec();
+
+    for (opening, &m_inv) in openings.iter().zip(m_invs) {
+        let packed_m_inv = F::Packing::from(m_inv);
+        let (packed_ys, sfx_ys) = F::Packing::pack_slice_with_suffix(inner_row);
+
+        if let Some(r_vertically_packed) = &opening.r_vertically_packed {
+            let uninit = qp_ys.spare_capacity_mut();
```

Comment:
this is ugly, but saves ~20% over `qp_ys.extend_from_slice(qp_.as_slice())`. Because that would have to go in the inner loop due to no const_generics, and it does a bounds check each time

Reply:
because there's no Packing::as_array, so we can't `qp_ys.extend(..flat_map(|| packed.as_slice()))`, that would return a reference